File: app/src/main/java/com/chaitupenjudcoder/recyclerviews/BucksTransactionsRecycler.java
package com.chaitupenjudcoder.recyclerviews;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.widget.Toast;

import com.chaitupenjudcoder.AddIncomeExpenseActivity;
import com.chaitupenjudcoder.buckstrack.R;
import com.chaitupenjudcoder.buckstrack.databinding.ItemSpendingListBinding;
import com.chaitupenjudcoder.datapojos.IncomeExpense;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

import static com.chaitupenjudcoder.BucksActivity.BUCKS_STRING_IS_INCOME_EXTRA;
import static com.chaitupenjudcoder.BucksActivity.INCOME_EXPENSE_OBJECT_EXTRA;

public class BucksTransactionsRecycler extends RecyclerView.Adapter<BucksTransactionsRecycler.BucksTransactionsAdapter> {

    private Context cxt;
    private ArrayList<IncomeExpense> allTransactions;

    public BucksTransactionsRecycler(Context cxt, ArrayList<IncomeExpense> allTransactions) {
        this.cxt = cxt;
        this.allTransactions = allTransactions;
    }

    public String convertDate(String dtFormat1, String dtFormat2, String dateStr) {
        SimpleDateFormat format1 = new SimpleDateFormat(dtFormat1, Locale.ENGLISH);
        SimpleDateFormat format2 = new SimpleDateFormat(dtFormat2, Locale.ENGLISH);
        Date date = null;
        try {
            date = format1.parse(dateStr);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return format2.format(date);
    }

    @NonNull
    @Override
    public BucksTransactionsAdapter onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        ItemSpendingListBinding items = ItemSpendingListBinding.inflate(inflater, viewGroup, false);
        return new BucksTransactionsAdapter(items);
    }

    @Override
    public void onBindViewHolder(@NonNull BucksTransactionsAdapter bucksTransactionsAdapter, int i) {
        bucksTransactionsAdapter.itemViewBinding.cvTransactions.setAnimation(AnimationUtils.loadAnimation(cxt, R.anim.transaction_card_scale_animation));
        IncomeExpense ie = allTransactions.get(i);
        ie.setDate(convertDate("dd-MM-yyyy", "dd/MM/yyyy", ie.getDate()));
        bucksTransactionsAdapter.bind(ie);
        bucksTransactionsAdapter.itemViewBinding.setTransactionClickListener(bucksTransactionsAdapter);
    }

    @Override
    public int getItemCount() {
        return allTransactions.size();
    }

    public class BucksTransactionsAdapter extends RecyclerView.ViewHolder {
        ItemSpendingListBinding itemViewBinding;

        BucksTransactionsAdapter(@NonNull ItemSpendingListBinding itemViewBinding) {
            super(itemViewBinding.getRoot());
            this.itemViewBinding = itemViewBinding;
        }

        public void bind(IncomeExpense ie) {
            this.itemViewBinding.setIncExpTransac(ie);
            this.itemViewBinding.executePendingBindings();
        }

        public void updateTransaction(IncomeExpense ie) {
            Intent in = new Intent(cxt, AddIncomeExpenseActivity.class);
            in.putExtra(BUCKS_STRING_IS_INCOME_EXTRA, ie.getBucksString().equals("income"));
            in.putExtra(INCOME_EXPENSE_OBJECT_EXTRA, ie);
            in.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            cxt.startActivity(in);
        }
    }
}

File: app/src/main/java/com/chaitupenjudcoder/recyclerviews/BucksTransactionsRecycler.kt
package com.chaitupenjudcoder.recyclerviews

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.recyclerview.widget.RecyclerView
import com.chaitupenjudcoder.AddIncomeExpenseActivity
import com.chaitupenjudcoder.BucksActivity
import com.chaitupenjudcoder.buckstrack.R
import com.chaitupenjudcoder.buckstrack.databinding.ItemSpendingListBinding
import com.chaitupenjudcoder.datapojos.IncomeExpense
import com.chaitupenjudcoder.firebasehelpers.SharedPreferencesHelper
import com.chaitupenjudcoder.recyclerviews.BucksTransactionsRecycler.BucksTransactionsAdapter
import java.util.*

class BucksTransactionsRecycler(
    private val cxt: Context,
    private val allTransactions: ArrayList<IncomeExpense>
) : RecyclerView.Adapter<BucksTransactionsAdapter>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): BucksTransactionsAdapter = BucksTransactionsAdapter(
            ItemSpendingListBinding.inflate(
                LayoutInflater.from(viewGroup.context),
                viewGroup,
                false
            )
        )

    override fun onBindViewHolder(bucksTransactionsAdapter: BucksTransactionsAdapter, i: Int) {
        bucksTransactionsAdapter.itemViewBinding.cvTransactions.animation =
            AnimationUtils.loadAnimation(cxt, R.anim.transaction_card_scale_animation)
        val ie = allTransactions[i]
        ie.setDateFormat(SharedPreferencesHelper(cxt).getDateFormatPref("dd-MM-yyyy"))
        bucksTransactionsAdapter.bind(ie)
        bucksTransactionsAdapter.itemViewBinding.transactionClickListener = bucksTransactionsAdapter
    }

    override fun getItemCount() = allTransactions.size

    inner class BucksTransactionsAdapter internal constructor(var itemViewBinding: ItemSpendingListBinding) :
        RecyclerView.ViewHolder(itemViewBinding.root) {

        fun bind(ie: IncomeExpense?) {
            itemViewBinding.apply {
                incExpTransac = ie

                executePendingBindings()
            }
        }

        fun updateTransaction(ie: IncomeExpense) {
            Intent(cxt, AddIncomeExpenseActivity::class.java).also { intent ->
                intent.putExtra(BucksActivity.BUCKS_STRING_IS_INCOME_EXTRA, ie.bucksString == "income")
                intent.putExtra(BucksActivity.INCOME_EXPENSE_OBJECT_EXTRA, ie)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                cxt.startActivity(intent)
            }
        }
    }
}